File: python/0645.py
# Link: https://leetcode.com/problems/set-mismatch/description/
# Runtime: 166 ms
# Memory: 18.56 MB


class Solution:
    def findErrorNums(self, nums: List[int]) -> List[int]:
        expected_sum = len(nums) * (len(nums) + 1) // 2
        actual_sum = sum(nums)

        different_numbers = set()
        duplicate = 0
        i, is_found = 0, False
        while (i < len(nums)) and (not is_found):
            if nums[i] in different_numbers:
                duplicate = nums[i]
                is_found = True

            different_numbers.add(nums[i])

            i += 1

        return [duplicate, expected_sum - (actual_sum - duplicate)]

File: c++/0645.cpp
// Link: https://leetcode.com/problems/set-mismatch/description/
// Runtime: 36 ms
// Memory: 32.22 MB

class Solution {
public:
  std::vector<int> findErrorNums(const std::vector<int> &nums) {
    int expectedSum = nums.size() * (nums.size() + 1) / 2;
    int actualSum = std::reduce(nums.begin(), nums.end());

    std::unordered_set<int> differentNumbers;
    int i = 0, duplicate = 0;
    bool isFound = false;
    while ((i < static_cast<int>(nums.size())) && (!isFound)) {
      if (differentNumbers.contains(nums[i])) {
        duplicate = nums[i];
        isFound = true;
      }

      differentNumbers.insert(nums[i]);

      i++;
    }

    return {duplicate, expectedSum - (actualSum - duplicate)};
  }
};
